### PR TITLE
Stop diffing blueprints and collections

### PR DESCRIPTION
**Pull request by andrewjstone**

### CODE REVIEW

**Comment by andrewjstone** [comment]
This is the implementation of the first checkbox in https://github.com/oxidecomputer/omicron/issues/7240

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 1]
```diff
@@ -2129,44 +2127,6 @@ pub mod test {
         assert_eq!(diff.sleds_modified.len(), 0);
     }
 
-    #[test]
-    fn test_initial() {
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:2133]: The only thing this test does is diff a blueprint and a collection, so it had to go.

File: nexus/reconfigurator/planning/src/planner.rs [hunk 2]
```diff
@@ -1831,7 +1831,9 @@ mod test {
         // NOTE: Expunging a disk doesn't immediately delete datasets; see the
         // "decommissioned_disk_cleaner" background task for more context.
         assert_eq!(diff.datasets.removed.len(), 0);
-        assert_eq!(diff.datasets.modified.len(), 0);
+
+        // The disposition has changed from `InService` to `Expunged`
```
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/planner.rs:1835]: This is actually a good change! The dataset was actually modified, we just couldn't see it with the reduction in form to the lowest common denominator shared with inventory that doesn't contain disposition.

File: nexus/types/src/deployment.rs [hunk 3]
```diff
@@ -1065,6 +983,7 @@ impl BlueprintDatasetDisposition {
 /// Information about a dataset as recorded in a blueprint
 #[derive(Debug, Clone, Eq, PartialEq, JsonSchema, Deserialize, Serialize)]
 pub struct BlueprintDatasetConfig {
+    // TODO: Display this in diffs - leave for now, for backwards compat
```
Comment (andrewjstone) [inline nexus/types/src/deployment.rs:986]: I'm trying to change as little of the diff output as possible with this change to demonstrate correctness.

In a follow up PR I'll add a column for disposition.

File: nexus/reconfigurator/planning/src/planner.rs (hunk 2 above)
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/planner.rs:1835]: Is it worth adding a check on the contents of `datasets.modified.first()` (mostly for clarity)? E.g., the checks a few lines below this that confirm the modified zone is crucible and it was expunged.

File: nexus/reconfigurator/planning/tests/output/planner_expunge_clickhouse_zones_after_policy_is_changed_3_4.txt [hunk 4]
```diff
@@ -121,31 +121,27 @@ to:   blueprint d895ef50-9978-454c-bdfb-b8dbe2c9a918
...
     dataset name                                                                                                     dataset uuid                           quota     reservation   compression
     -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-    oxp_5fe54077-c016-49a9-becb-14993f133d43/crucible                                                                2a41b700-5494-4aa1-900b-20618125c1d0   none      none          off        
-    oxp_51c788ff-de33-43f7-b9c5-f5f56bf80736/crucible                                                                0bbf7ff9-f6a6-49d8-8dc4-c75b2fdb5531   none      none          off        
+    oxp_f52832ea-60d7-443b-9847-df5384bfc8e2/crucible                                                                cc43cb24-2e55-415d-9a2a-49a6b924b284   none      none          off        
+    oxp_9825ff38-f07d-44a1-9efc-55a25e72015b/crucible                                                                34de0f9d-ab3a-4047-bbf3-d2bf4beb2995   none      none          off
```
Comment (jgallagher) [inline nexus/reconfigurator/planning/tests/output/planner_expunge_clickhouse_zones_after_policy_is_changed_3_4.txt:125]: I tried to look at why this output changed. It looks like this list is sorted _only_ by kind; within one kind, there doesn't appear to be any sorting. Maybe we can make this more stable - if this analysis looks right, maybe this fixes it?

```diff
--- a/nexus/types/src/deployment/blueprint_diff.rs
+++ b/nexus/types/src/deployment/blueprint_diff.rs
@@ -562,7 +562,7 @@ impl BpTableData for DiffDatasetsDetails {
         // by sorting by dataset kind: after UUID redaction, that produces
         // a stable table ordering for datasets.
         let mut rows = self.datasets.values().collect::<Vec<_>>();
-        rows.sort_unstable_by_key(|d| &d.kind);
+        rows.sort_unstable_by_key(|d| (&d.kind, &d.name));
         rows.into_iter().map(move |dataset| {
             BpTableRow::from_strings(state, dataset.as_strings())
         })

**Review by jgallagher — APPROVED** [review]

File: nexus/reconfigurator/planning/tests/output/planner_expunge_clickhouse_zones_after_policy_is_changed_3_4.txt (hunk 4 above)
Comment (andrewjstone) [reply nexus/reconfigurator/planning/tests/output/planner_expunge_clickhouse_zones_after_policy_is_changed_3_4.txt:125]: I think this is a good change. It's going to result in a bit of churn in other output files though. I'd suggest looking at the coming commit standalone.

Comment (jgallagher) [reply nexus/reconfigurator/planning/tests/output/planner_expunge_clickhouse_zones_after_policy_is_changed_3_4.txt]: LGTM :+1:

File: nexus/reconfigurator/planning/src/planner.rs (hunk 2 above)
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/planner.rs]: Great idea! Added in [14bc84c](https://github.com/oxidecomputer/omicron/pull/7252/commits/14bc84c28574b646dfb053f9bd0f6bf464830b80)

**Pull request merged**